Add the ability to store multipart/alternative messages

settermjd opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: default X.Y.z branch or the oldest support X.Y.z
  * Bugfix: default X.Y.z branch or the oldest support X.Y.z
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: default X.Y.z branch or the oldest support X.Y.z
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | **yes**
| Bugfix        | **yes**
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description

This change adds the ability to properly parse messages where the parent message's content-type is set to `multipart/alternative` into multiple Part objects. Previously, this content-type header would be ignored, leaving alternative parts being rolled into one single Part object. See [7.2.3 The Multipart/alternative subtype](https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html) for more information.

At the moment, I don't know if this will cause an effective BC break.

settermjd committed:
Add the ability to store multipart/alternative messages

This change adds the ability to properly parse messages where the
content type is "multipart/alternative" into multiple Part objects.
Previously, the "multipart/alternative" content-type header would be
ignored, leaving alternative parts being rolled into one larger Part
object.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Update documentation

This is a small change to correct some minor points in the related
documentation.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Use an existing constant instead of adding a new one

I wasn't aware at the time that I created the previous commit with the
constant, of the existence of a constant with the same value. So this
commit removes my duplicate constant, replacing it with the existing
one.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

weierophinney commented on src/Part.php:
```diff
@@ -64,6 +64,33 @@ class Part
+     * Stores a list of sub parts of this part
+     *
+     * It supports RFC 1341 7.2.3: "The Multipart/alternative subtype".
+     * While it can a list of parts without limitation, the intention,
```
```suggestion
     * While it can store a list of parts without limitation, the intention,
```

weierophinney commented on test/MessageTest.php:
```diff
@@ -221,4 +224,177 @@ public function testFromStringWithCrlfAndRfc2822FoldedHeaders()
+
+<div dir=3D"ltr">This is a test email with 1 attachment.<br clear=3D"all"><=
+div><br></div>-- <br><div class=3D"gmail_signature" data-smartmail=3D"gmail=
+_signature"><div dir=3D"ltr"><img src=3D"https://sendgrid.com/brand/sg-logo=
```
Use an example.com domain instead to keep it agnostic and to prevent somebody from copying and sending this.

settermjd replied:
Works for me. Will do.

weierophinney reviewed (changes requested):
A couple minor change suggestions, but overall looks solid (though I might prefer we typehint the `$parts` property as `Part[]`).

Also... Run phpcbf and phpcs before your next push.

settermjd committed:
Update src/Part.php

Co-authored-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Fix type hinting of the parts property

As per @mwop's suggestion, this change properly typehints the protected
$parts property in Part to return an array of Part objects.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Fix up code styling

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Add further assertions when testing multipart alternative messages

This change adds a series of assertions for attachments on
multipart/alternative messages, to help verify that they contain what
they're expected to contain. I felt that it was helpful to have
assertions for attachments (or non text/HTML content) as well, so that
more of the code could be properly tested.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Remove domain-specific references

These were originally used because it was a fairly generic email that I
had at the time, while writing the tests. However, as per @mwop's
suggestion, it should be more generic (and not preference a given,
commercial, email provider).

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd commented:
All changes made and checks pass, except for Autocloser.

settermjd committed:
Set a Part filename if the part stores an attachment

This change simplistically sets a Part's filename property if the
content disposition header is set, it's an attachment, and the filename
property is set. I've deliberately avoided any considerations about
sanitising the filename, deciding, at this stage, to leave that to the
calling code up the line.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Correct a small bug in setting an attachment's filename

Previously, the code only set a filename for a message part if the part
contained a content-disposition header where the disposition type was
set to "attachment" and the filename parameter was set. After re-reading
https://www.w3.org/Protocols/HTTP/Issues/content-disposition.txt, I see
that this was wrong, as the content-disposition header can have the
filename parameter set if the type is attachment or inline.

This commit corrects that earlier mistake.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

gsteel commented:
This library is being abandoned: https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2023-12-04-TSC-Minutes.md
